### PR TITLE
[corlib] CoreRT System.Threading.Tasks

### PR DESCRIPTION
**Pull request by EgorBo**
Replaces `System.Threading.Tasks` implementation with corefx+corert.
WIP
Requires changes in the corert and corefx submodules to compile.

Implements #6967

### CODE REVIEW

**Comment by EgorBo** [comment]
net_4_x_corlib_xunit-test  Total: 12474, Errors: 0, Failed: 1, Skipped: 2, Time: 15.794s

File: mcs/class/corlib/Makefile [hunk 1]
```diff
@@ -30,6 +30,7 @@ RESX_RESOURCE_STRING = \
 	../../../external/corefx/src/System.Collections.Concurrent/src/Resources/Strings.resx \
 	../../../external/corefx/src/System.Memory/src/Resources/Strings.resx \
 	../../../external/corefx/src/System.Runtime.InteropServices.RuntimeInformation/src/Resources/Strings.resx \
+	../../../external/corefx/src/System.Threading.Tasks.Parallel/src/Resources/Strings.resx \
```
Comment (marek-safar) [inline mcs/class/corlib/Makefile]: Why do we need this?

File: mcs/class/corlib/System/Delegate.cs [hunk 2]
```diff
@@ -115,6 +115,10 @@ protected Delegate (Type target, string method)
 			}
 		}
 
+		internal IntPtr GetNativeFunctionPointer() {
+			throw new NotImplementedException (); // or return method_ptr?
```
Comment (marek-safar) [inline mcs/class/corlib/System/Delegate.cs]: Where is it used?

File: mcs/class/corlib/corert/RuntimeThread.cs [hunk 3]
```diff
@@ -1,14 +1,37 @@
+using System;
+using System.Threading;
+
 namespace Internal.Runtime.Augments
 {
-	class RuntimeThread
+	internal class RuntimeThread
```
Comment (marek-safar) [inline mcs/class/corlib/corert/RuntimeThread.cs]: internal is redundant

File: mcs/class/corlib/corert/RuntimeThread.cs [hunk 4]
```diff
@@ -1,14 +1,37 @@
...
 	{
-		public static RuntimeThread InitializeThreadPoolThread ()
+		Thread thread = null;
```
Comment (marek-safar) [inline mcs/class/corlib/corert/RuntimeThread.cs]: should be readonly and extra null initialization

File: mcs/class/corlib/corlib.dll.sources [hunk 5]
```diff
@@ -1594,21 +1594,31 @@ ReferenceSources/AppContextDefaultValues.cs
...
+../../../external/corert/src/System.Private.Interop/src/InteropExtensions/Lock.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/LockHolder.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/TaskScheduler.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Future.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/TaskContinuation.cs
+../../../external/corert/src/System.Private.CoreLib/shared/System/Threading/Tasks/TaskCompletionSource.cs
+../../../external/corert/src/System.Private.CoreLib/shared/System/Threading/Tasks/TaskSchedulerException.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/ThreadPoolTaskScheduler.cs
+../../../external/corefx/src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/*.cs
```
Comment (marek-safar) [inline mcs/class/corlib/corlib.dll.sources]: I don't like this, why do we need to pull whole System.Threading.Tasks.Parallel

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 6]
```diff
@@ -1107,8 +1107,8 @@ private bool ValidHandle()
             // This will result in a "leak" of sorts (since the handle will not be cleaned up)
             // but the process is exiting anyway.
             //
-            // During AD-unload, we don�t finalize live objects until all threads have been 
-            // aborted out of the AD.  Since these locked regions are CERs, we won�t abort them 
+            // During AD-unload, we don’t finalize live objects until all threads have been
```
Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Undo this change

File: mcs/class/corlib/corert/TaskScheduler.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+namespace System.Threading.Tasks
+{
+	partial class TaskScheduler
+	{
+		internal void InternalQueueTask(Task task)
+		{
+			throw new NotImplementedException ();
```
Comment (marek-safar) [inline mcs/class/corlib/corert/TaskScheduler.cs]: Is this correct?

File: mcs/class/corlib/corert/Task.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
+using System.Collections.Generic;
```
Comment (marek-safar) [inline mcs/class/corlib/corert/Task.cs]: Where did you copy this from?

File: mcs/class/corlib/corlib.dll.sources (hunk 5 above)
Comment (EgorBo) [reply mcs/class/corlib/corlib.dll.sources]: @marek-safar  Um but what is wrong with it? there are just a few files under this wildcard and I replaced old Parallel we had with corert's

File: mcs/class/corlib/Makefile (hunk 1 above)
Comment (EgorBo) [reply mcs/class/corlib/Makefile]: Because Parallel from corefx uses some string resources

File: mcs/class/corlib/corert/Task.cs (hunk 8 above)
Comment (EgorBo) [reply mcs/class/corlib/corert/Task.cs]: @marek-safar https://github.com/EgorBo/mono/blob/master/mcs/class/referencesource/mscorlib/system/threading/Tasks/Task.cs#L220-L246

File: mcs/class/corlib/corert/TaskScheduler.cs (hunk 7 above)
Comment (EgorBo) [reply mcs/class/corlib/corert/TaskScheduler.cs]: It's a temp stub as the PR is still WIP

File: mcs/class/corlib/corlib.dll.sources [hunk 9]
```diff
@@ -1594,21 +1594,31 @@ ReferenceSources/AppContextDefaultValues.cs
 ../referencesource/mscorlib/system/threading/waithandle.cs
 ../referencesource/mscorlib/system/threading/waithandleExtensions.cs
 
+../../../external/corefx/src/Common/src/System/Threading/Tasks/RendezvousAwaitable.cs
+../../../external/corert/src/Common/src/System/Collections/Generic/LowLevelDictionary.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/DebuggerSupport.cs
+../../../external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/DebuggerSupport.Dummy.cs
+../../../external/corefx/src/Common/src/System/Threading/Tasks/BeginEndAwaitableAdapter.cs
+../../../external/corert/src/System.Private.CoreLib/shared/System/Threading/Tasks/TaskCanceledException.cs
```
Comment (marek-safar) [inline mcs/class/corlib/corlib.dll.sources]: As discussed get the shared bits from `external/corefx/src/Common/src/CoreLib/` and not corert

File: mcs/class/corlib/corert/Task.cs (hunk 8 above)
Comment (marek-safar) [reply mcs/class/corlib/corert/Task.cs]: Add this commend to the file with link who needs that as we should be able to remove it later

File: mcs/class/corlib/corert/AwaitTaskContinuation.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
+namespace System.Threading.Tasks
```
Comment (marek-safar) [inline mcs/class/corlib/corert/AwaitTaskContinuation.cs]: should be under corefx/

File: mcs/class/corlib/corert/SynchronizationContext.cs [hunk 11]
```diff
@@ -0,0 +1,7 @@
+namespace System.Threading
```
Comment (marek-safar) [inline mcs/class/corlib/corert/SynchronizationContext.cs]: should be under corefx/

File: mcs/class/corlib/corlib.dll.sources (hunk 5 above)
Comment (marek-safar) [reply mcs/class/corlib/corlib.dll.sources]: My error, I exchange it with some other parts

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 6 above)
Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: Please undo the change

**Comment by marek-safar** [comment]
We need to hide the types in `Internal` namespaces
Also remove redundant TypeForwarderTo

File: mcs/class/corlib/Test/System.Threading.Tasks/TaskFactoryTest.cs [hunk 12]
```diff
@@ -586,8 +585,9 @@ public void FromAsync_BeginCallback ()
 					Assert.IsFalse (Thread.CurrentThread.IsThreadPoolThread, "#12");
 
 					called2 = true;
-					b.Invoke (null);
-					return null;
+					var ar = Task.CompletedTask;
+					b.Invoke (ar);
```
Comment (EgorBo) [inline mcs/class/corlib/Test/System.Threading.Tasks/TaskFactoryTest.cs]: b.Invoke(null) throws NRE on .NET Core and .NET Framework so I fixed it this way

**Comment by marek-safar** [comment]
@EgorBo API changes looks fine but I think we should hide or possibly not even include `DependencyReductionRootAttribute`

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 13]
```diff
@@ -1107,8 +1107,8 @@ private bool ValidHandle()
             // This will result in a "leak" of sorts (since the handle will not be cleaned up)
             // but the process is exiting anyway.
             //
-            // During AD-unload, we don�t finalize live objects until all threads have been 
-            // aborted out of the AD.  Since these locked regions are CERs, we won�t abort them 
+            // During AD-unload, we don�t finalize live objects until all threads have been 
+            // aborted out of the AD.  Since these locked regions are CERs, we won�t abort them
```
Comment (ForNeVeR) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: It already was bad enough, but not it seems even worse! 😄 

(could you fix the unicode here, please?)

**Comment by marek-safar** [comment]
@monojenkins commit apidiff
